calico-vpp-agent/services/service_server.go

```diff
@@ -325,48 +348,43 @@ func (s *Server) configureSnat() (err error) {
 	return nil
 }
 
-func (s *Server) findMatchingService(ep *v1.Endpoints) *v1.Service {
-	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(ep)
-	if err != nil {
-		s.log.Errorf("Error getting endpoint %+v key: %v", ep, err)
-		return nil
-	}
-	value, found, err := s.serviceStore.GetByKey(key)
-	if err != nil {
-		s.log.Errorf("Error getting service %s: %v", key, err)
-		return nil
-	}
-	if !found {
-		s.log.Debugf("Service %s not found", key)
+func (s *Server) findMatchingService(ep *discoveryv1.EndpointSlice) *v1.Service {
+	labelSvcName, ok := ep.Labels["kubernetes.io/service-name"]
+	if !ok {
+		s.log.Debugf("No label for endpointslice %s", ep.Name)
 		return nil
 	}
-	service, ok := value.(*v1.Service)
-	if !ok {
-		panic("s.serviceStore.GetByKey did not return value of type *v1.Service")
+	services := s.serviceStore.List()
+	for _, ele := range services {
+		svc, ok := ele.(*v1.Service)
+		if !ok {
+			panic("s.serviceStore.List() did not return value of type *v1.Service")
+		}
+		if labelSvcName == svc.Name && ep.Namespace == svc.Namespace {
+			return svc
+		}
 	}
-	return service
+	s.log.Debugf("No service found for endpointslice %s", ep.Name)
+	return nil
 }
 
-func (s *Server) findMatchingEndpoint(service *v1.Service) *v1.Endpoints {
-	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(service)
-	if err != nil {
-		s.log.Errorf("Error getting service %+v key: %v", service, err)
-		return nil
-	}
-	value, found, err := s.endpointStore.GetByKey(key)
-	if err != nil {
-		s.log.Errorf("Error getting endpoint %s: %v", key, err)
-		return nil
-	}
-	if !found {
-		s.log.Debugf("Endpoint %s not found", key)
-		return nil
-	}
-	endpoints, ok := value.(*v1.Endpoints)
-	if !ok {
-		panic("s.serviceStore.GetByKey did not return value of type *v1.Service")
+func (s *Server) findMatchingEndpoint(service *v1.Service) []*discoveryv1.EndpointSlice {
+	epslices := []*discoveryv1.EndpointSlice{}
+	endpointslices := s.endpointStore.List()
```

Comment:
Here I think we should store the endpoint slices by service name in the 
same fashion as Calico does it here [0] to avoid the walk.

[0] https://github.com/projectcalico/calico/blob/master/felix/serviceindex/serviceindex.go#L38